pkg/validation/attributes/attributes_parent.go

```diff
@@ -0,0 +1,48 @@
+package attributes
+
+import (
+	"github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
+	apiAttributes "github.com/devfile/api/v2/pkg/attributes"
+)
+
+// ValidateAndReplaceForParent validates the parent data for global attribute references(except parent overrides) and replaces them with the attribute value
+func ValidateAndReplaceForParent(attributes apiAttributes.Attributes, parent *v1alpha2.Parent) error {
+	var err error
+
+	if parent != nil {
+		switch {
+		case parent.Id != "":
+			// Validate parent id
+			if parent.Id, err = validateAndReplaceDataWithAttribute(parent.Id, attributes); err != nil {
+				return err
+			}
+		case parent.Uri != "":
+			// Validate parent uri
+			if parent.Uri, err = validateAndReplaceDataWithAttribute(parent.Uri, attributes); err != nil {
+				return err
+			}
+		case parent.Kubernetes != nil:
+			// Validate parent kubernetes name
+			if parent.Kubernetes.Name, err = validateAndReplaceDataWithAttribute(parent.Kubernetes.Name, attributes); err != nil {
+				return err
+			}
+
+			// Validate parent kubernetes namespace
+			if parent.Kubernetes.Namespace, err = validateAndReplaceDataWithAttribute(parent.Kubernetes.Namespace, attributes); err != nil {
+				return err
+			}
+		}
+
+		// Validate parent registry url
+		if parent.RegistryUrl, err = validateAndReplaceDataWithAttribute(parent.RegistryUrl, attributes); err != nil {
+			return err
+		}
+
+		// Note: No need to substitute parent overrides at this point. Call global attribute validation/substitution
+		// after merging the flattened parent devfile to the main devfile. Parent's global attribute key can
+		// be overridden in parent overrides or the alternative is to mention the attribute as a main devfile
+		// global attribute if parent devfile does not have a global attribute
```

Comment:
This is kind of subtle and results in a fairly complicated replacement flow:
1. Do attribute replacement to fill placeholders in `parent` section
    - This *might fail* if a component in the current devfile expects to use an attribute from its parent
2. Flatten parent into Devfile to get a devfile without parent elements
3. Do attribute replacement again, to handle placeholders in parent
4. Validate devfile only at this point

Global attributes + replacement means that there are various places where devfiles are invalid until some other steps happen.

Reply:
So I've thought about this when drafting this PR and the way I see it is (copying devfile snipppet from your comment below),

```yaml
metadata:
  name: parent-devfile
attributes:
  parent-attribute: "test"
components:
  - name: parent-component
    container:
      image: quay.io/test/image
---
metadata:
  name: main-devfile
attributes:
  parent-devfile-id: "parent-devfile"
parent:
  id: "{{parent-devfile-id}}"
components:
  - name: child-component
    image: quay.io/{{parent-attribute}}/image
```

To resolve this flow, in order

1. resolve the parent in the main-devfile, so in this case the id `{{parent-devfile-id}}`(uri/kubernetes/registryURL) needs to be resolved to fetch the parent
2. overwrite the parent-devfile with main-devfile's parent overrides (teh regular override convention)
3. merge parent-devfile to main-devfile (the regular merging convention)
4. now validate the flatten devfile for global attributes

So we definitely need to resolve a parent source in the main-devfile to fetch the parent and is the only requirement here, unless we say parent source cannot have a global attribute as well.. WDYT?

As for your comment regd various places in devfile maybe invalid until some steps happen - isn't that true for other devfile objects? For example, my main-devfile can be the following and is invalid by itself in the main-devfile but is valid in a flattened devfile.. 🤔 

```yaml
metadata:
  name: parent-devfile
components:
  - name: parent-component
    container:
      image: quay.io/test/image
---
metadata:
  name: main-devfile
parent:
  id: "parent-devfile-id"
command:
  - id: child-component
     exec:
        component: parent-component
```

Reply:
This makes sense to me (and good point about unflattened devfiles in general). I'm on the side of "no attributes in parent" except maybe in overrides to make step 1 simpler.

Reply:
@amisevsk Ack! Will remove global attributes from parent source but the overrides can have them(following the agreed upon observation that override object's id/name cannot have global attributes) and they will be resolved in the flattened devfile which is the step 4 mentioned above.

Reply:
Addressed in new commit :)